### DISCUSSION TITLE
Allow top-down frontend customisation

### DISCUSSION THREAD

**Discussion by ghost** [directus/directus#4522] [Feature Requests]
Right now you can customise the frontend bottom-up. So starting from interfaces/displays, and if the modification is larger, a new layout, and if even larger, a module, and if even larger a whole new frontend app.

This is a suggestion to allow top-down customization too, allowing the developer to use existing layouts/modules as part of a custom app. This can be done by allowing the developer to write custom private or public views. 

A good first step to that is to register the built-in modules, layouts and [the private view components](https://github.com/directus/directus/tree/main/app/src/views/private/components). This will allow building custom views with whatever branding aspects are needed as part of a new custom module, then simply plugging in existing components/layouts/modules.

This would also close #4204.

**Comment by rijkvanzanten** [comment]
I'm not a 100% sure what you mean by 

> A good first step to that is to register the built-in modules, layouts and the private view components.

Are you saying one would replace individual "core" components in order to customize the app? 🤔

**Reply by ghost** [reply]
Yeah - by registering those private view components, it becomes easier to rebuild the default `<private-view>` (and same for public views), this way anyone could quickly recreate a new view or overall "structure" for the app to fit their design/branding language while simply plugging-in the default existing modules and layouts.  Plus, right now you can't reference the built in layouts/modules (since they're not registered), so to build my own view I'd also have to rebuild all the default modules/layouts.

**Reply by rijkvanzanten** [reply]
Ahh gotcha

> Plus, right now you can't reference the built in layouts/modules (since they're not registered), so to build my own view I'd also have to rebuild all the default modules/layouts.

This is only partially true. The [views (private/public) are registered](https://github.com/directus/directus/blob/42507ae52f03423321b117f57fcb7374266cdb65/app/src/views/register.ts#L6-L7), as are all [layouts](https://github.com/directus/directus/blob/42507ae52f03423321b117f57fcb7374266cdb65/app/src/layouts/register.ts#L37), [displays](https://github.com/directus/directus/blob/42507ae52f03423321b117f57fcb7374266cdb65/app/src/displays/register.ts#L39) and [interfaces](https://github.com/directus/directus/blob/42507ae52f03423321b117f57fcb7374266cdb65/app/src/interfaces/register.ts#L38).
